<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>ATAC-Seq with Bioconductor • AtacSeqWorkshop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="ATAC-Seq with Bioconductor">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">AtacSeqWorkshop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Workflow.html">ATAC-Seq with Bioconductor</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>ATAC-Seq with Bioconductor</h1>
                        <h4 class="author">Sean Davis</h4>
            <address class="author_afil">
      National Cancer Institute<br><a class="author_email" href="mailto:#"></a><a href="mailto:seandavi@gmail.com">seandavi@gmail.com</a>
      </address>
                  
      
      
      <div class="hidden name"><code>Workflow.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      This workshop employs Bioconductor GenomicRanges and GenomicFeatures infrastructure to perform basic import of aligned reads followed by quality control of an ATAC-Seq dataset. It will also examine how to use fragment length (insert size) to isolate chromatin compartments of interest such as putative nucleosome-free regions. Moving on to a multi-sample analysis, peak regions (called separately) will inform similarities between samples. Statistical identification of differential accessibility will provide a second measure of sample differences and serve as potential regions of interest for enrichment analysis and biological hypothesis generation. Data visualization throughout using the Integrated Genome Viewer (IGV) can make findings more intuitive, so students are encouraged to export datasets from R in formats compatible with IGV.
    </div>
    
<div id="overview" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h1>
<div id="pre-requisites" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#pre-requisites" class="anchor"></a>Pre-requisites</h2>
<p>This workshop assumes:</p>
<ul>
<li>A working and up-to-date version of R</li>
<li>Basic knowledge of R syntax</li>
<li>Familiarity with the <em><a href="https://bioconductor.org/packages/3.10/GenomicRanges">GenomicRanges</a></em> package and range manipulations</li>
<li>Familiarity with BAM files and their contents</li>
</ul>
</div>
<div id="participation" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#participation" class="anchor"></a>Participation</h2>
<p>After a very brief review of ATAC-Seq and chromatin accessibility, students will work independently to follow this workflow. Additional materials are provided as links at the end of the workshop for those wanting deeper exposure. Additional materials include alignment from FASTQ files and peak calling.</p>
</div>
<div id="r-bioconductor-packages-used" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#r-bioconductor-packages-used" class="anchor"></a><em>R</em> / <em>Bioconductor</em> packages used</h2>
<ul>
<li>
<em><a href="https://bioconductor.org/packages/3.10/Rsamtools">Rsamtools</a></em><br>
</li>
<li>
<em><a href="https://bioconductor.org/packages/3.10/GenomicRanges">GenomicRanges</a></em><br>
</li>
<li>
<em><a href="https://bioconductor.org/packages/3.10/GenomicFeatures">GenomicFeatures</a></em><br>
</li>
<li><em><a href="https://bioconductor.org/packages/3.10/GenomicAlignments">GenomicAlignments</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.10/rtracklayer">rtracklayer</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.10/heatmaps">heatmaps</a></em></li>
</ul>
</div>
<div id="time-outline" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#time-outline" class="anchor"></a>Time outline</h2>
<p>An example for a 45-minute workshop:</p>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Introduction</td>
<td>15m</td>
</tr>
<tr class="even">
<td>Independent work</td>
<td>2-3hr</td>
</tr>
<tr class="odd">
<td>Additional exercises (optional, external)</td>
<td>up to 12 hours</td>
</tr>
</tbody>
</table>
</div>
<div id="learning-goals" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#learning-goals" class="anchor"></a>Learning goals</h2>
<ul>
<li>Describe how to import sequence alignments in BAM format into R</li>
<li>Relate fragment size to genomic characteristics such as nucleosome occupancy and open chromatin.</li>
<li>Perform basic alignment manipulations in R to enrich ATAC-seq data for chromatin characteristics.</li>
<li>Gain familiarity with the IGV genome browser and examining data in genomic context.</li>
<li>Visualize summaries of genomic signal using profile plots and heatmaps.</li>
</ul>
</div>
<div id="learning-objectives" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#learning-objectives" class="anchor"></a>Learning objectives</h2>
<ul>
<li>Load and save genomic data in BAM and BigWig formats [GenomicAlignments and rtracklayer].</li>
<li>Perform basic QC plots from ATAC-Seq data.</li>
<li>Isolate nucleosome-free and mononucleosome regions from ATAC-seq data.</li>
<li>Install and use IGV to visualize data in genomic context.</li>
<li>Create profile plots using the heatmaps package.</li>
</ul>
</div>
</div>
<div id="background" class="section level1">
<h1 class="hasAnchor">
<a href="#background" class="anchor"></a>Background</h1>
<p>Chromatin accessibility assays measure the extent to which DNA is open and accessible. Such assays now use high throughput sequencing as a quantitative readout. DNAse assays, first using microarrays<span class="citation">(Crawford, Davis, et al. <a href="#ref-Crawford2006-nt">2006</a>)</span> and then DNAse-Seq <span class="citation">(Crawford, Holt, et al. <a href="#ref-Crawford2006-bk">2006</a>)</span>, requires a larger amount of DNA and is labor-indensive and has been largely supplanted by ATAC-Seq <span class="citation">(Buenrostro et al. <a href="#ref-Buenrostro2013-dz">2013</a>)</span>.</p>
<p>The Assay for Transposase Accessible Chromatin with high-throughput sequencing (ATAC-seq) method maps chromatin accessibility genome-wide. This method quantifies DNA accessibility with a hyperactive Tn5 transposase that cuts and inserts sequencing adapters into regions of chromatin that are accessible. High throughput sequencing of fragments produced by the process map to regions of increased accessibility, transcription factor binding sites, and nucleosome positioning. The method is both fast and sensitive and can be used as a replacement for DNAse <em>and</em> MNase. A schematic of the protocol is given in Figure @ref(fig:protocol).</p>
<div class="figure">
<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/d/d9/ATAC-Seq_Fig.1.pdf/page1-984px-ATAC-Seq_Fig.1.pdf.jpg" alt="Schematic overview of ATAC-Seq protocol. Figure from Wikipedia."><p class="caption">Schematic overview of ATAC-Seq protocol. Figure from Wikipedia.</p>
</div>
<p>An early review of chromatin accessibility assays <span class="citation">(Tsompana and Buck <a href="#ref-Tsompana2014-yh">2014</a>)</span> compares the use cases, pros and cons, and expected signals from each of the most common approaches (Figure @ref(fig:chromatinAssays)).</p>
<div class="figure">
<img src="imgs/chromatin_accessibility_buck.png" alt="Chromatin accessibility methods, compared. Representative DNA fragments generated by each assay are shown, with end locations within chromatin defined by colored arrows. Bar diagrams represent data signal obtained from each assay across the entire region. The footprint created by a transcription factor (TF) is shown for ATAC-seq and DNase-seq experiments." width="808"><p class="caption">
Chromatin accessibility methods, compared. Representative DNA fragments generated by each assay are shown, with end locations within chromatin defined by colored arrows. Bar diagrams represent data signal obtained from each assay across the entire region. The footprint created by a transcription factor (TF) is shown for ATAC-seq and DNase-seq experiments.
</p>
</div>
<p>The first manuscript describing ATAC-Seq protocol and findings outlined how ATAC-Seq data “line up” with other datatypes such as ChIP-seq and DNAse-seq (Figure @ref(fig:greenleaf)). They also highlight how fragment length correlates with specific genomic regions and characteristics <span class="citation">(Buenrostro et al. <a href="#ref-Buenrostro2013-dz">2013</a>, Figure 3)</span>.</p>
<div class="figure">
<img src="imgs/greenleaf_chip_dnase_atac.png" alt="Multimodal chromatin comparisons. From [@Buenrostro2013-dz], Figure 4. (a) CTCF footprints observed in ATAC-seq and DNase-seq data, at a specific locus on chr1. (b) Aggregate ATAC-seq footprint for CTCF (motif shown) generated over binding sites within the genome (c) CTCF predicted binding probability inferred from ATAC-seq data, position weight matrix (PWM) scores for the CTCF motif, and evolutionary conservation (PhyloP). Right-most column is the CTCF ChIP-seq data (ENCODE) for this GM12878 cell line, demonstrating high concordance with predicted binding probability." width="836"><p class="caption">
Multimodal chromatin comparisons. From <span class="citation">(Buenrostro et al. <a href="#ref-Buenrostro2013-dz">2013</a>)</span>, Figure 4. (a) CTCF footprints observed in ATAC-seq and DNase-seq data, at a specific locus on chr1. (b) Aggregate ATAC-seq footprint for CTCF (motif shown) generated over binding sites within the genome (c) CTCF predicted binding probability inferred from ATAC-seq data, position weight matrix (PWM) scores for the CTCF motif, and evolutionary conservation (PhyloP). Right-most column is the CTCF ChIP-seq data (ENCODE) for this GM12878 cell line, demonstrating high concordance with predicted binding probability.
</p>
</div>
<p>Buenrostro et al. provide a detailed protocol for performing ATAC-Seq and quality control of results <span class="citation">(Buenrostro et al. <a href="#ref-Buenrostro2015-dz">2015</a>)</span>. Updated and modified protocols that improve on signal-to-noise and reduce input DNA requirements have been described.</p>
<div id="informatics-overview" class="section level2">
<h2 class="hasAnchor">
<a href="#informatics-overview" class="anchor"></a>Informatics overview</h2>
<p>ATAC-Seq protocols typically utilize paired-end sequencing protocols. The reads are aligned to the respective <em>genome</em> using <code>bowtie2</code>, <code>BWA</code>, or other short-read aligner. The result, after appropriate manipulation, often using <code>samtools</code>, results in a BAM file. Among other details, the BAM format includes columns for:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1">knitr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/knitr/topics/include_graphics">include_graphics</a></span>(<span class="st">'imgs/bam_shot.png'</span>)</a></code></pre></div>
<div class="figure">
<img src="imgs/bam_shot.png" alt="A BAM file in text form. The output of `samtools view` is the text format of the BAM file (called SAM format). Bioconductor and many other tools use BAM files for input. Note that BAM files also often include an index `.bai` file that enables random access into the file; one can read just a genomic region without having to read the entire file." width="1126"><p class="caption">
A BAM file in text form. The output of <code>samtools view</code> is the text format of the BAM file (called SAM format). Bioconductor and many other tools use BAM files for input. Note that BAM files also often include an index <code>.bai</code> file that enables random access into the file; one can read just a genomic region without having to read the entire file.
</p>
</div>
<ul>
<li>sequence name (<code>chr1</code>)</li>
<li>start position (integer)</li>
<li>a <em>CIGAR</em> string that describes the alignment in a compact form</li>
<li>the sequence to which the pair aligns</li>
<li>the position to which the pair aligns</li>
<li>a bit flag field that describes multiple characteristics of the alignment</li>
<li>the sequence and quality string of the read</li>
<li>additional tags that tend to be aligner-specific</li>
</ul>
<p>Duplicate fragments (those with the <em>same</em> start and end position of other reads) are marked and likely discarded. Reads that fail to align “properly” are also often excluded from analysis. It is worth noting that most software packages allow simple “marking” of such reads and that there is usually no need to create a special BAM file before proceeding with downstream work.</p>
<p>After alignment and BAM processing, the workflow can switch to <em>Bioconductor</em>.</p>
</div>
<div id="working-with-sequencing-data-in-bioconductor" class="section level2">
<h2 class="hasAnchor">
<a href="#working-with-sequencing-data-in-bioconductor" class="anchor"></a>Working with sequencing data in Bioconductor</h2>
<p>The <em>Bioconductor</em> project includes several infrastructure packages for dealing with ranges (sequence name, start, end, +/- strand) on sequences <span class="citation">(Lawrence et al. <a href="#ref-Lawrence2013-he">2013</a>)</span> as well as capabilities with working with Fastq files directly <span class="citation">(Morgan et al. <a href="#ref-morgan2016rsamtools">2016</a>)</span>.</p>
<table class="table">
<caption>Commonly used Bioconductor and their high-level use cases.</caption>
<colgroup>
<col width="50%">
<col width="50%">
</colgroup>
<thead><tr class="header">
<th>Package</th>
<th>Use cases</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><em><a href="https://bioconductor.org/packages/3.10/Rsamtools">Rsamtools</a></em></td>
<td>low level access to FASTQ, VCF, SAM, BAM, BCF formats</td>
</tr>
<tr class="even">
<td><em><a href="https://bioconductor.org/packages/3.10/GenomicRanges">GenomicRanges</a></em></td>
<td>Container and methods for handling genomic reagions</td>
</tr>
<tr class="odd">
<td><em><a href="https://bioconductor.org/packages/3.10/GenomicFeatures">GenomicFeatures</a></em></td>
<td>Work with transcript databases, gff, gtf and BED formats</td>
</tr>
<tr class="even">
<td><em><a href="https://bioconductor.org/packages/3.10/GenomicAlignments">GenomicAlignments</a></em></td>
<td>Reader for BAM format</td>
</tr>
<tr class="odd">
<td><em><a href="https://bioconductor.org/packages/3.10/rtracklayer">rtracklayer</a></em></td>
<td>import and export multiple UCSC file formats including BigWig and Bed</td>
</tr>
</tbody>
</table>
<p>As noted in the previous section, the output of an ATAC-Seq experiment is a BAM file. As paired-end sequencing is a commonly-applied approach for ATAC-Seq, the <code>readGAlignmentPairs</code> function is the appropriate method to use.</p>
</div>
</div>
<div id="data-import-and-quality-control" class="section level1">
<h1 class="hasAnchor">
<a href="#data-import-and-quality-control" class="anchor"></a>Data import and quality control</h1>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(GenomicAlignments)</a></code></pre></div>
<p>Reading a paired-end BAM file looks a bit complicated, but the following code will:</p>
<ol style="list-style-type: decimal">
<li>Read the included BAM file.</li>
<li>Include read pairs only (<code>isPaired = TRUE</code>)</li>
<li>Include properly paired reads (<code>isProperPair = TRUE</code>)</li>
<li>Include reads with mapping quality &gt;= 1</li>
<li>Add a couple of additional fields, <code>mapq</code> (mapping quality) and <code>isize</code> (insert size) to the default fields.</li>
</ol>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">greenleaf =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/GenomicAlignments/topics/readGAlignments">readGAlignmentPairs</a></span>(</a>
<a class="sourceLine" id="cb3-2" data-line-number="2">  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="dt">package=</span><span class="st">'AtacSeqWorkshop'</span>,</a>
<a class="sourceLine" id="cb3-3" data-line-number="3">              <span class="st">'extdata/Sorted_ATAC_21_22.bam'</span>),</a>
<a class="sourceLine" id="cb3-4" data-line-number="4">  <span class="dt">param =</span> <span class="kw">ScanBamParam</span>(</a>
<a class="sourceLine" id="cb3-5" data-line-number="5">    <span class="dt">mapqFilter =</span> <span class="dv">1</span>, </a>
<a class="sourceLine" id="cb3-6" data-line-number="6">    <span class="dt">flag =</span> <span class="kw">scanBamFlag</span>(</a>
<a class="sourceLine" id="cb3-7" data-line-number="7">      <span class="dt">isPaired =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb3-8" data-line-number="8">      <span class="dt">isProperPair =</span> <span class="ot">TRUE</span>), </a>
<a class="sourceLine" id="cb3-9" data-line-number="9">    <span class="dt">what =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"mapq"</span>, <span class="st">"isize"</span>)))</a></code></pre></div>
<p><em>Exercise</em>: What is the class of <code>greenleaf</code>? <em>Exercise</em>: Use the <code><a href="https://www.rdocumentation.org/packages/GenomicAlignments/topics/GAlignmentPairs-class">GenomicAlignments::first()</a></code> accessor to get the first read of the pair as a <code>GAlignments</code> object. Save the result as a variable called <code>gl_first_read</code>. Use the <code>mcols</code> accessor to find the “metadata columns” of <code>gl_first_read</code>. <em>Exercise</em>: How many read pairs map to each chromosome?</p>
<p>We can make plot of the number of reads mapping to each chromosome.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(ggplot2)</a>
<a class="sourceLine" id="cb4-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(dplyr)</a>
<a class="sourceLine" id="cb4-3" data-line-number="3">chromCounts =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/table">table</a></span>(<span class="kw">seqnames</span>(greenleaf)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb4-4" data-line-number="4"><span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/data.frame">data.frame</a></span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb4-5" data-line-number="5"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">rename</a></span>(<span class="dt">chromosome=</span>Var1, <span class="dt">count =</span> Freq)</a></code></pre></div>
<p>To keep things small, the example BAM file includes only chromosomes 21 and 22.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(chromCounts, <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x=</span>chromosome, <span class="dt">y=</span>count)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-2" data-line-number="2"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span>(<span class="dt">stat=</span><span class="st">'identity'</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb5-3" data-line-number="3"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="dt">axis.text.x =</span> <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(<span class="dt">angle =</span> <span class="dv">45</span>, <span class="dt">hjust =</span> <span class="dv">1</span>))</a></code></pre></div>
<div class="figure">
<img src="Workflow_files/figure-html/plotChromCounts-1.png" alt="Reads per chromosome. In our example data, we are using only chromosomes 21 and 22." width="700"><p class="caption">
Reads per chromosome. In our example data, we are using only chromosomes 21 and 22.
</p>
</div>
<p>Normalizing by the chromosome length can yield the reads per megabase which should crudely be similar across all chromosomes.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1">chromCounts =<span class="st"> </span>chromCounts <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">readsPerMb =</span> (count<span class="op">/</span>(<span class="kw">seqlengths</span>(greenleaf)<span class="op">/</span><span class="fl">1e6</span>)))</a></code></pre></div>
<p>And show a plot. For two chromosomes, this is a little underwhelming.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(chromCounts, <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x=</span>chromosome, <span class="dt">y=</span>readsPerMb)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb7-2" data-line-number="2"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span>(<span class="dt">stat=</span><span class="st">'identity'</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb7-3" data-line-number="3"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="dt">axis.text.x =</span> <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(<span class="dt">angle =</span> <span class="dv">45</span>, <span class="dt">hjust =</span> <span class="dv">1</span>)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb7-4" data-line-number="4"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>()</a></code></pre></div>
<div class="figure">
<img src="Workflow_files/figure-html/plotReadProps-1.png" alt="Read counts normalized by chromosome length. This is not a particularly important plot, but it can be useful to see the relative contribution of each chromosome given its length." width="700"><p class="caption">
Read counts normalized by chromosome length. This is not a particularly important plot, but it can be useful to see the relative contribution of each chromosome given its length.
</p>
</div>
<div id="coverage" class="section level2">
<h2 class="hasAnchor">
<a href="#coverage" class="anchor"></a>Coverage</h2>
<p>The <code>coverage</code> method for genomic ranges calculates, for each base, the number of overlapping features. In the case of a BAM file from ATAC-Seq converted to a GAlignmentPairs object, the coverage gives us an idea of the extent to which reads pile up to form peaks.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1">cvg =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/GenomicAlignments/topics/coverage-methods">coverage</a></span>(greenleaf)</a>
<a class="sourceLine" id="cb8-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/class">class</a></span>(cvg)</a></code></pre></div>
<pre><code>## [1] "SimpleRleList"
## attr(,"package")
## [1] "IRanges"</code></pre>
<p>The coverage is returned as a <code>SimpleRleList</code> object. Using <code>names</code> can get us the names of the elements of the list.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(cvg)</a></code></pre></div>
<pre><code>##  [1] "chr1"  "chr2"  "chr3"  "chr4"  "chr5"  "chr6"  "chr7"  "chr8" 
##  [9] "chr9"  "chr10" "chr11" "chr12" "chr13" "chr14" "chr15" "chr16"
## [17] "chr17" "chr18" "chr19" "chr20" "chr21" "chr22" "chrX"  "chrY" 
## [25] "chrM"</code></pre>
<p>There is a name for each chromosome. Looking at the <code>chr21</code> entry:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1">cvg<span class="op">$</span>chr21</a></code></pre></div>
<pre><code>## integer-Rle of length 48129895 with 397462 runs
##   Lengths: 9411376      50      11      50 ...      14      28   10806
##   Values :       0       2       0       2 ...       2       1       0</code></pre>
<p>we see that each chromosome is represented as an <code>Rle</code>, short for run-length-encoding. Simply put, since along the chromosome there are many repeated values, we can recode the long vector as a set of (length: value) pairs. For example, if the first 9,410,000 base pairs have <code>0</code> coverage, we encode that as (9,410,000: 0). Doing that across the chromosome can very significantly reduce the memory use for genomic coverage.</p>
<p>The following little function, <code>plotCvgHistByChrom</code> can plot a histogram of the coverage for a chromosome.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1">plotCvgHistByChrom =<span class="st"> </span><span class="cf">function</span>(cvg, chromosome) {</a>
<a class="sourceLine" id="cb14-2" data-line-number="2">  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(ggplot2)</a>
<a class="sourceLine" id="cb14-3" data-line-number="3">  cvgcounts =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/as.data.frame">as.data.frame</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/table">table</a></span>(cvg[[chromosome]]))</a>
<a class="sourceLine" id="cb14-4" data-line-number="4">  cvgcounts[,<span class="dv">1</span>] =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/numeric">as.numeric</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/character">as.character</a></span>(cvgcounts[,<span class="dv">1</span>]))</a>
<a class="sourceLine" id="cb14-5" data-line-number="5">  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(cvgcounts) =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">'Coverage'</span>, <span class="st">'Count'</span>)</a>
<a class="sourceLine" id="cb14-6" data-line-number="6">  <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(cvgcounts, <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x=</span>Coverage, <span class="dt">y=</span>Count)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb14-7" data-line-number="7"><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="st">"Chromosome"</span>,chromosome)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb14-8" data-line-number="8"><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>(<span class="dt">alpha=</span><span class="fl">0.5</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb14-9" data-line-number="9"><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html">geom_smooth</a></span>(<span class="dt">span=</span><span class="fl">0.2</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb14-10" data-line-number="10"><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_log10</a></span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb14-11" data-line-number="11"><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() </a>
<a class="sourceLine" id="cb14-12" data-line-number="12">}</a>
<a class="sourceLine" id="cb14-13" data-line-number="13"><span class="cf">for</span>(i <span class="cf">in</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">'chr21'</span>, <span class="st">'chr22'</span>)) {</a>
<a class="sourceLine" id="cb14-14" data-line-number="14">  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(<span class="kw">plotCvgHistByChrom</span>(cvg, i))</a>
<a class="sourceLine" id="cb14-15" data-line-number="15">}</a></code></pre></div>
<p><img src="Workflow_files/figure-html/unnamed-chunk-7-1.png" width="700"><img src="Workflow_files/figure-html/unnamed-chunk-7-2.png" width="700"></p>
</div>
<div id="fragment-lengths" class="section level2">
<h2 class="hasAnchor">
<a href="#fragment-lengths" class="anchor"></a>Fragment Lengths</h2>
<p>The first ATAC-Seq manuscript <span class="citation">(Buenrostro et al. <a href="#ref-Buenrostro2013-dz">2013</a>)</span> highlighted the relationship between fragment length and nucleosomes (see Figure @ref{fig:flgreenleaf}).</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1">knitr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/knitr/topics/include_graphics">include_graphics</a></span>(<span class="st">'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3959825/bin/nihms554473f2.jpg'</span>)</a></code></pre></div>
<div class="figure">
<img src="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3959825/bin/nihms554473f2.jpg" alt="Relationship between fragment length and nucleosome number."><p class="caption">Relationship between fragment length and nucleosome number.</p>
</div>
<p>Remember that we loaded the example BAM file with insert sizes (<code>isize</code>). We can use that “column” to examine the fragment lengths (another name for insert size). Also, note that the insert size for the <code>first</code> read and the <code>second</code> are the same (absolute value). Here, we will use <code>first</code>.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1">GenomicAlignments<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/GenomicAlignments/topics/GAlignmentPairs-class">first</a></span>(greenleaf)</a>
<a class="sourceLine" id="cb16-2" data-line-number="2"><span class="kw">mcols</span>(GenomicAlignments<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/GenomicAlignments/topics/GAlignmentPairs-class">first</a></span>(greenleaf))</a>
<a class="sourceLine" id="cb16-3" data-line-number="3"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/class">class</a></span>(<span class="kw">mcols</span>(GenomicAlignments<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/GenomicAlignments/topics/GAlignmentPairs-class">first</a></span>(greenleaf)))</a>
<a class="sourceLine" id="cb16-4" data-line-number="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(<span class="kw">mcols</span>(GenomicAlignments<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/GenomicAlignments/topics/GAlignmentPairs-class">first</a></span>(greenleaf))<span class="op">$</span>isize)</a>
<a class="sourceLine" id="cb16-5" data-line-number="5">fraglengths =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/MathFun">abs</a></span>(<span class="kw">mcols</span>(GenomicAlignments<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/GenomicAlignments/topics/GAlignmentPairs-class">first</a></span>(greenleaf))<span class="op">$</span>isize)</a></code></pre></div>
<p>We can plot the fragment length density (histogram) using the <code>density</code> function.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot">plot</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/density">density</a></span>(fraglengths, <span class="dt">bw=</span><span class="fl">0.05</span>), <span class="dt">xlim=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">0</span>,<span class="dv">1000</span>))</a></code></pre></div>
<div class="figure">
<img src="Workflow_files/figure-html/fraglength-1.png" alt="Fragment length histogram. " width="700"><p class="caption">
Fragment length histogram.
</p>
</div>
<p><em>Exercise</em>: Adjust the <code>xlim</code>, <code>bw</code>, and try <code>log="y"</code> in the plot to highlight features present in figure .</p>
<p>And for fun, the ggplot2 version:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(dplyr)</a>
<a class="sourceLine" id="cb18-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(ggplot2)</a>
<a class="sourceLine" id="cb18-3" data-line-number="3">fragLenPlot &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/table">table</a></span>(fraglengths) <span class="op">%&gt;%</span><span class="st"> </span>data.frame <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">rename</a></span>(<span class="dt">InsertSize =</span> fraglengths, </a>
<a class="sourceLine" id="cb18-4" data-line-number="4">    <span class="dt">Count =</span> Freq) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">InsertSize =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/numeric">as.numeric</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/vector">as.vector</a></span>(InsertSize)), </a>
<a class="sourceLine" id="cb18-5" data-line-number="5">    <span class="dt">Count =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/numeric">as.numeric</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/vector">as.vector</a></span>(Count))) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> InsertSize, <span class="dt">y =</span> Count)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb18-6" data-line-number="6"><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span>()</a>
<a class="sourceLine" id="cb18-7" data-line-number="7"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(fragLenPlot <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/lims.html">lims</a></span>(<span class="dt">x=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="op">-</span><span class="dv">1</span>,<span class="dv">250</span>)))</a></code></pre></div>
<p><img src="Workflow_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>Knowing that the nucleosome-free regions will have insert sizes shorter than one nucleosome, we can isolate the read pairs that have that characteristic.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1">gl_nf =<span class="st"> </span>greenleaf[<span class="kw">mcols</span>(GenomicAlignments<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/GenomicAlignments/topics/GAlignmentPairs-class">first</a></span>(greenleaf))<span class="op">$</span>isize<span class="op">&lt;</span><span class="dv">100</span>]</a></code></pre></div>
<p>And the mononucleosome reads will be between 187 and 250 base pairs for insert size/fragment length.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1">gl_mn =<span class="st"> </span>greenleaf[<span class="kw">mcols</span>(GenomicAlignments<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/GenomicAlignments/topics/GAlignmentPairs-class">first</a></span>(greenleaf))<span class="op">$</span>isize<span class="op">&gt;</span><span class="dv">187</span> <span class="op">&amp;</span><span class="st"> </span></a>
<a class="sourceLine" id="cb20-2" data-line-number="2"><span class="st">                  </span><span class="kw">mcols</span>(GenomicAlignments<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/GenomicAlignments/topics/GAlignmentPairs-class">first</a></span>(greenleaf))<span class="op">$</span>isize<span class="op">&lt;</span><span class="dv">250</span>  ]</a></code></pre></div>
<p>Finally, we expect nucleosome-free reads to be enriched near the TSS while mononucleosome reads should not be. We will use the <em><a href="https://bioconductor.org/packages/3.10/heatmaps">heatmaps</a></em> package to take a look at these two sets of reads with respect to the tss of the human genome.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(TxDb.Hsapiens.UCSC.hg19.knownGene)</a>
<a class="sourceLine" id="cb21-2" data-line-number="2">proms =<span class="st"> </span><span class="kw">promoters</span>(TxDb.Hsapiens.UCSC.hg19.knownGene, <span class="dv">250</span>, <span class="dv">250</span>)</a>
<a class="sourceLine" id="cb21-3" data-line-number="3">seqs =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">'chr21'</span>,<span class="st">'chr22'</span>)</a>
<a class="sourceLine" id="cb21-4" data-line-number="4"><span class="kw">seqlevels</span>(proms, <span class="dt">pruning.mode=</span><span class="st">'coarse'</span>) =<span class="st"> </span>seqs <span class="co"># only chromosome 21 and 22</span></a></code></pre></div>
<p>Take a look at the <em><a href="https://bioconductor.org/packages/3.10/heatmaps">heatmaps</a></em> package vignette to learn more about the heatmaps package capabilities.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(heatmaps)</a>
<a class="sourceLine" id="cb22-2" data-line-number="2">gl_nf_hm =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/heatmaps/topics/CoverageHeatmap">CoverageHeatmap</a></span>(proms,<span class="kw"><a href="https://www.rdocumentation.org/packages/GenomicAlignments/topics/coverage-methods">coverage</a></span>(gl_nf),<span class="dt">coords=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="op">-</span><span class="dv">250</span>,<span class="dv">250</span>))</a>
<a class="sourceLine" id="cb22-3" data-line-number="3"><span class="kw"><a href="https://www.rdocumentation.org/packages/heatmaps/topics/label">label</a></span>(gl_nf_hm) =<span class="st"> "NucFree"</span></a>
<a class="sourceLine" id="cb22-4" data-line-number="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/heatmaps/topics/scale">scale</a></span>(gl_nf_hm)=<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">0</span>,<span class="dv">10</span>)</a>
<a class="sourceLine" id="cb22-5" data-line-number="5"><span class="kw"><a href="https://www.rdocumentation.org/packages/heatmaps/topics/plotHeatmapMeta">plotHeatmapMeta</a></span>(gl_nf_hm)</a></code></pre></div>
<div class="figure">
<img src="Workflow_files/figure-html/unnamed-chunk-13-1.png" alt="Enrichment of nucleosome free reads just upstream of the TSS." width="700"><p class="caption">
Enrichment of nucleosome free reads just upstream of the TSS.
</p>
</div>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" data-line-number="1">gl_mn_hm =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/heatmaps/topics/CoverageHeatmap">CoverageHeatmap</a></span>(proms,<span class="kw"><a href="https://www.rdocumentation.org/packages/GenomicAlignments/topics/coverage-methods">coverage</a></span>(gl_mn),<span class="dt">coords=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="op">-</span><span class="dv">250</span>,<span class="dv">250</span>))</a>
<a class="sourceLine" id="cb23-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/heatmaps/topics/label">label</a></span>(gl_mn_hm) =<span class="st"> "MonoNuc"</span></a>
<a class="sourceLine" id="cb23-3" data-line-number="3"><span class="kw"><a href="https://www.rdocumentation.org/packages/heatmaps/topics/scale">scale</a></span>(gl_mn_hm)=<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">0</span>,<span class="dv">10</span>)</a>
<a class="sourceLine" id="cb23-4" data-line-number="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/heatmaps/topics/plotHeatmapMeta">plotHeatmapMeta</a></span>(gl_mn_hm)</a></code></pre></div>
<div class="figure">
<img src="Workflow_files/figure-html/unnamed-chunk-14-1.png" alt="Depletion of nucleosome free reads just upstream of the TSS." width="700"><p class="caption">
Depletion of nucleosome free reads just upstream of the TSS.
</p>
</div>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/heatmaps/topics/plotHeatmapList">plotHeatmapList</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(gl_mn_hm, gl_nf_hm))</a></code></pre></div>
<div class="figure">
<img src="Workflow_files/figure-html/unnamed-chunk-15-1.png" alt="Comparison of signals at TSS. Mononucleosome data on the left, nucleosome-free on the right." width="700"><p class="caption">
Comparison of signals at TSS. Mononucleosome data on the left, nucleosome-free on the right.
</p>
</div>
</div>
</div>
<div id="viewing-data-in-igv" class="section level1">
<h1 class="hasAnchor">
<a href="#viewing-data-in-igv" class="anchor"></a>Viewing data in IGV</h1>
<p>Install IGV from <a href="https://software.broadinstitute.org/software/igv/download">here</a>.</p>
<p>We export the greenleaf data as a BigWig file.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb25-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(rtracklayer)</a>
<a class="sourceLine" id="cb25-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/rtracklayer/topics/BigWigFile">export.bw</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/GenomicAlignments/topics/coverage-methods">coverage</a></span>(greenleaf),<span class="st">'greenleaf.bw'</span>)</a></code></pre></div>
<p><em>Exercise</em>: In IGV, choose <code>hg19</code>. Then, load the <code>greenleaf.bw</code> file and explore chromosomes 21 and 22. <em>Exercise</em>: Export the nucleosome-free portion of the data as a BigWig file and examine that in IGV. Where do you expect to see the strongest signals?</p>
</div>
<div id="additional-work" class="section level1">
<h1 class="hasAnchor">
<a href="#additional-work" class="anchor"></a>Additional work</h1>
<p>For those working extensively on ATAC-Seq, there is a great workflow/tutorial available from Thomas Carrol:</p>
<p><a href="https://rockefelleruniversity.github.io/RU_ATAC_Workshop.html" class="uri">https://rockefelleruniversity.github.io/RU_ATAC_Workshop.html</a></p>
<p>Feel free to work through it. In addition to the work above, there is also the <em><a href="https://bioconductor.org/packages/3.10/ATACseqQC">ATACseqQC</a></em> package vignette that offers more than just QC. At least a couple more packages are available in <em>Bioconductor</em>.</p>
</div>
<div id="appendix" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#appendix" class="anchor"></a>Appendix</h1>
<div id="session-info" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h2>
<pre><code>## R Under development (unstable) (2019-01-14 r75992)
## Platform: x86_64-apple-darwin15.6.0 (64-bit)
## Running under: macOS Mojave 10.14.2
## 
## Matrix products: default
## BLAS: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] rtracklayer_1.45.1                     
##  [2] heatmaps_1.9.0                         
##  [3] TxDb.Hsapiens.UCSC.hg19.knownGene_3.2.2
##  [4] GenomicFeatures_1.37.3                 
##  [5] AnnotationDbi_1.47.0                   
##  [6] dplyr_0.8.1                            
##  [7] ggplot2_3.2.0                          
##  [8] GenomicAlignments_1.21.4               
##  [9] Rsamtools_2.1.2                        
## [10] Biostrings_2.53.0                      
## [11] XVector_0.25.0                         
## [12] SummarizedExperiment_1.15.5            
## [13] DelayedArray_0.11.2                    
## [14] BiocParallel_1.19.0                    
## [15] matrixStats_0.54.0                     
## [16] Biobase_2.45.0                         
## [17] GenomicRanges_1.37.14                  
## [18] GenomeInfoDb_1.21.1                    
## [19] IRanges_2.19.10                        
## [20] S4Vectors_0.23.17                      
## [21] BiocGenerics_0.31.4                    
## [22] knitr_1.23                             
## [23] BiocStyle_2.13.2                       
## 
## loaded via a namespace (and not attached):
##  [1] bitops_1.0-6           fs_1.3.1               EBImage_4.27.0        
##  [4] bit64_0.9-7            RColorBrewer_1.1-2     progress_1.2.2        
##  [7] httr_1.4.0             rprojroot_1.3-2        tools_3.6.0           
## [10] backports_1.1.4        R6_2.4.0               KernSmooth_2.23-15    
## [13] DBI_1.0.0              lazyeval_0.2.2         colorspace_1.4-1      
## [16] withr_2.1.2            tidyselect_0.2.5       prettyunits_1.0.2     
## [19] bit_1.1-14             compiler_3.6.0         xml2_1.2.0            
## [22] desc_1.2.0             labeling_0.3           bookdown_0.11         
## [25] scales_1.0.0           pkgdown_1.3.0          commonmark_1.7        
## [28] stringr_1.4.0          digest_0.6.19          tiff_0.1-5            
## [31] fftwtools_0.9-8        rmarkdown_1.13         jpeg_0.1-8            
## [34] pkgconfig_2.0.2        htmltools_0.3.6        plotrix_3.7-6         
## [37] highr_0.8              htmlwidgets_1.3        rlang_0.4.0           
## [40] rstudioapi_0.10        RSQLite_2.1.1          RCurl_1.95-4.12       
## [43] magrittr_1.5           GenomeInfoDbData_1.2.1 Matrix_1.2-17         
## [46] Rcpp_1.0.1             munsell_0.5.0          abind_1.4-5           
## [49] stringi_1.4.3          yaml_2.2.0             MASS_7.3-51.4         
## [52] zlibbioc_1.31.0        grid_3.6.0             blob_1.1.1            
## [55] crayon_1.3.4           lattice_0.20-38        hms_0.4.2             
## [58] locfit_1.5-9.1         pillar_1.4.2           codetools_0.2-16      
## [61] biomaRt_2.41.3         XML_3.98-1.20          glue_1.3.1            
## [64] evaluate_0.14          BiocManager_1.30.4     png_0.1-7             
## [67] gtable_0.3.0           purrr_0.3.2            assertthat_0.2.1      
## [70] xfun_0.8               roxygen2_6.1.1         tibble_2.1.3          
## [73] memoise_1.1.0</code></pre>
</div>
<div id="macs2" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#macs2" class="anchor"></a>MACS2</h2>
<p>The MACS2 package is a commonly-used package for calling peaks. Installation and other details are available<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>.</p>
<pre><code>pip install macs2</code></pre>
</div>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-Buenrostro2013-dz">
<p>Buenrostro, Jason D, Paul G Giresi, Lisa C Zaba, Howard Y Chang, and William J Greenleaf. 2013. “Transposition of Native Chromatin for Fast and Sensitive Epigenomic Profiling of Open Chromatin, DNA-binding Proteins and Nucleosome Position.” <em>Nature Methods</em> 10 (12):1213–8. <a href="https://doi.org/10.1038/nmeth.2688" class="uri">https://doi.org/10.1038/nmeth.2688</a>.</p>
</div>
<div id="ref-Buenrostro2015-dz">
<p>Buenrostro, Jason D, Beijing Wu, Howard Y Chang, and William J Greenleaf. 2015. “ATAC-seq: A Method for Assaying Chromatin Accessibility Genome-Wide.” <em>Current Protocols in Molecular Biology / Edited by Frederick M. Ausubel ... [et Al.]</em> 109 (January):21.29.1–9. <a href="https://doi.org/10.1002/0471142727.mb2129s109" class="uri">https://doi.org/10.1002/0471142727.mb2129s109</a>.</p>
</div>
<div id="ref-Crawford2006-nt">
<p>Crawford, Gregory E, Sean Davis, Peter C Scacheri, Gabriel Renaud, Mohamad J Halawi, Michael R Erdos, Roland Green, Paul S Meltzer, Tyra G Wolfsberg, and Francis S Collins. 2006. “DNase-chip: A High-Resolution Method to Identify DNase I Hypersensitive Sites Using Tiled Microarrays.” <em>Nature Methods</em> 3 (7):503–9. <a href="http://www.ncbi.nlm.nih.gov/pubmed/16791207?dopt=AbstractPlus" class="uri">http://www.ncbi.nlm.nih.gov/pubmed/16791207?dopt=AbstractPlus</a>.</p>
</div>
<div id="ref-Crawford2006-bk">
<p>Crawford, Gregory E, Ingeborg E Holt, James Whittle, Bryn D Webb, Denise Tai, Sean Davis, Elliott H Margulies, et al. 2006. “Genome-Wide Mapping of DNase Hypersensitive Sites Using Massively Parallel Signature Sequencing (MPSS).” <em>Genome Research</em> 16 (1):123–31. <a href="http://www.ncbi.nlm.nih.gov/pubmed/16344561?dopt=AbstractPlus" class="uri">http://www.ncbi.nlm.nih.gov/pubmed/16344561?dopt=AbstractPlus</a>.</p>
</div>
<div id="ref-Lawrence2013-he">
<p>Lawrence, Michael, Wolfgang Huber, Hervé Pagès, Patrick Aboyoun, Marc Carlson, Robert Gentleman, Martin T Morgan, and Vincent J Carey. 2013. “Software for Computing and Annotating Genomic Ranges.” <em>PLoS Computational Biology</em> 9 (8):e1003118. <a href="https://doi.org/10.1371/journal.pcbi.1003118" class="uri">https://doi.org/10.1371/journal.pcbi.1003118</a>.</p>
</div>
<div id="ref-morgan2016rsamtools">
<p>Morgan, Martin, Herve Pages, V Obenchain, and N Hayden. 2016. “Rsamtools: Binary Alignment (Bam), Fasta, Variant Call (Bcf), and Tabix File Import.” <em>R Package Version</em> 1 (0):677–89.</p>
</div>
<div id="ref-Tsompana2014-yh">
<p>Tsompana, Maria, and Michael J Buck. 2014. “Chromatin Accessibility: A Window into the Genome.” <em>Epigenetics &amp; Chromatin</em> 7 (1):33. <a href="https://doi.org/10.1186/1756-8935-7-33" class="uri">https://doi.org/10.1186/1756-8935-7-33</a>.</p>
</div>
</div>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p><a href="https://github.com/taoliu/MACS" class="uri">https://github.com/taoliu/MACS</a><a href="#fnref1" class="footnote-back">↩</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#overview">Overview</a><ul class="nav nav-pills nav-stacked">
<li><a href="#pre-requisites">Pre-requisites</a></li>
      <li><a href="#participation">Participation</a></li>
      <li><a href="#r-bioconductor-packages-used"><em>R</em> / <em>Bioconductor</em> packages used</a></li>
      <li><a href="#time-outline">Time outline</a></li>
      <li><a href="#learning-goals">Learning goals</a></li>
      <li><a href="#learning-objectives">Learning objectives</a></li>
      </ul>
</li>
      <li>
<a href="#background">Background</a><ul class="nav nav-pills nav-stacked">
<li><a href="#informatics-overview">Informatics overview</a></li>
      <li><a href="#working-with-sequencing-data-in-bioconductor">Working with sequencing data in Bioconductor</a></li>
      </ul>
</li>
      <li>
<a href="#data-import-and-quality-control">Data import and quality control</a><ul class="nav nav-pills nav-stacked">
<li><a href="#coverage">Coverage</a></li>
      <li><a href="#fragment-lengths">Fragment Lengths</a></li>
      </ul>
</li>
      <li><a href="#viewing-data-in-igv">Viewing data in IGV</a></li>
      <li><a href="#additional-work">Additional work</a></li>
      <li>
<a href="#appendix">Appendix</a><ul class="nav nav-pills nav-stacked">
<li><a href="#session-info">Session info</a></li>
      <li><a href="#macs2">MACS2</a></li>
      </ul>
</li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Sean Davis.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
